gh-116968: Reimplement Tier 2 counters

gvanrossum opened the pull request:
This introduces a unified 16-bit backoff counter type (`_Py_BackoffCounter`), shared between the Tier 1 adaptive specialization machinery and the Tier 2 optimizer.

The latter's side exit temperature now uses exponential backoff, and starts initially at 64, to avoid creating side exit traces for code that hasn't been respecialized yet (since the latter only happens after the cooldown counter has reached zero from an initial value of 52).

The threshold value for back-edge optimizations is no longer dynamic; we just use a backoff counter initialized to (16, 4).

<!-- gh-issue-number: gh-116968 -->
* Issue: gh-116968
<!-- /gh-issue-number -->

gvanrossum committed:
Baby steps: reimplement thresholds using adaptive counter abstractions

gvanrossum committed:
Make temperature an adaptive counter like the rest

gvanrossum committed:
Fix tests

gvanrossum committed:
Remove dead adaptive_counter_jump_init()

gvanrossum committed:
Fix no-GIL build failure in _COLD_EXIT

gvanrossum committed:
Use the right named constant in initial temperature

gvanrossum commented on Python/bytecodes.c:
```diff
@@ -4159,34 +4157,36 @@ dummy_func(
-                        Py_DECREF(previous);
-                        tstate->previous_executor = Py_None;
-                        GOTO_UNWIND();
+            #if ENABLE_SPECIALIZATION
```
Reviewers: this file is better viewed with the "hide whitespace" option, because I indented a huge block to rearrange things to move the counter manipulation into the `#if ENABLE_SPECIALIZATION` while keeping `GOTO_TIER_ONE(target);` out of it.

gvanrossum replied:
I dedented the code again after realizing it shouldn't be in `#if ENABLE_SPECIALIZATION`, so please disregard this suggestion.
`

gvanrossum committed:
Add pycore_backoff.h, and include it, but don't use it yet

gvanrossum committed:
Reimplement adaptive counters in terms of backoff_counter

gvanrossum committed:
Redefine T2 temperature as a backoff counter

gvanrossum committed:
Don't increment branch cache (bitmask) in INSTRUMENTED_INSTRUCTION

gvanrossum committed:
Don't update unreachable counters

gvanrossum committed:
Don't increment branch cache (bitmask) in INSTRUMENTED_LINE

gvanrossum committed:
Simplify dynamic counter initialization for JUMP_BACKWARD

gvanrossum committed:
Revert "Don't increment branch cache (bitmask) in INSTRUMENTED_LINE"

This reverts commit 8f79a6061b78f2a03913578c94ec3896eb43b700.

gvanrossum committed:
Different approach to avoid incrementing bitmask in INSTRUMENTED_LINE

gvanrossum committed:
Different approach to avoid incrementing bitmask in INSTRUMENTED_INSTRUCTION

gvanrossum committed:
Fix dynamic counter initialization for JUMP_BACKWARD

gvanrossum committed:
Get rid of (unused) resume_threshold

gvanrossum committed:
Hopeful fix for non-clang builds

`((backoff_counter_t){.counter = UNREACHABLE_BACKOFF})`
is not a compile-time constant.

gvanrossum commented on Include/internal/pycore_backoff.h:
```diff
@@ -0,0 +1,117 @@
+increment_backoff_counter(backoff_counter_t counter)
+{
+    if (!is_unreachable_backoff_counter(counter)) {
+        assert(counter.value != 0xFFF);
```
This assert fails in the free-threading build. The old code didn't have such an assert in `INCREMENT_ADAPTIVE_COUNTER().` But doesn't that imply that incrementing the counter could overflow? And couldn't that jeopardize the use of `INCREMENT_ADAPTIVE_COUNTER()`? Its only use is to prevent the very next instruction being specialized, which apparently breaks instrumentation. @markshannon Can you help me understand how this is supposed to work, and help me find an alternative approach?

FWIW adding the assert to `INCREMENT_ADAPTIVE_COUNTER()` causes the same crash there in no-GIL mode, so I think that, at least in no-GIL mode, this really goes wrong somehow.

gvanrossum replied:
Hm. In no-GIL mode, `DECREMENT_ADAPTIVE_COUNTER()` is never called, so we might as well make `INCREF_ADAPTIVE_COUNTER()` a dummy, which makes the tests pass.

But that leaves me with the question, could we reach `INCREMENT_ADAPTIVE_COUNTER()` immediately after the counter has been reset to `(0xFFF, 12)`? In that case, the assert will fail -- but then again, without the assert, the counter will end up being set to zero, having the opposite effect of what's intended: the very next instruction will be specialized. If this scenario is possible, maybe we should use some other macro, e.g. one that does `counter |= 0x1F`.

gvanrossum replied:
Solved by a "pause" operation which just does `.value |= 1`. (**CORRECTED** to `|=`)

gvanrossum commented on Python/bytecodes.c:
```diff
@@ -3973,7 +3968,12 @@ dummy_func(
+                    next_opcode != POP_JUMP_IF_NOT_NONE &&
+                    next_opcode != POP_JUMP_IF_NONE) {
+                    INCREMENT_ADAPTIVE_COUNTER(next_instr[1].cache);
+                }
```
@markshannon I had to add this because the naive increment would occasionally mess with the bitmap of recent branches used by the `POP_JUMP_IF` opcodes, and I had added an assert that complains if `backoff > 12`. I guess it was previously also messing with the bitmap, but nobody cared, and the effect is rarely significant. But this also makes me wonder if we don't need a different primitive than `INCREMENT_ADAPTIVE_COUNTER()`.

markshannon replied:
It might be simpler to unconditionally set the counter/bit-vector to a value > 1.
Messing up the branch profiling won't matter for `INSTRUMENTED_INSTRUCTION`. Performance will be terrible anyway.

gvanrossum replied:
Okay, I'll just do `value |= 1`.

gvanrossum commented on Python/ceval.c:
```diff
@@ -818,6 +819,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+            assert(original_opcode != POP_JUMP_IF_FALSE);
+            assert(original_opcode != POP_JUMP_IF_TRUE);
+            assert(original_opcode != POP_JUMP_IF_NONE);
+            assert(original_opcode != POP_JUMP_IF_NOT_NONE);
```
@markshannon Same thing here.

gvanrossum replied:
(I think -- it doesn't actually fail without this, AFAICT, but it seems to make sense.)

gvanrossum replied:
The `|= 1` will work here too.

gvanrossum committed:
In no-GIL mode, make INCREMENT_ADAPTIVE_COUNTER() a no-op

markshannon commented on Include/internal/pycore_backoff.h:
```diff
@@ -0,0 +1,117 @@
+
+typedef struct {
+    union {
+        uint16_t counter;
```
Is looks like this is just used for `is_unreachable_backoff_counter`. Maybe have `backoff == 15 && value == MAX` indicate that it is unreachable?

gvanrossum replied:
It's also used to get the full 16-bit value as an int. And I like that the unreachable value cannot be constructed from a (value, backoff) pair because the backoff needs to be <= 12.

gvanrossum replied:
(I had to break down and remove the `assert(backoff <= 12)` because there were corner cases where the "unreachable" value would be put back in line this. But I still feel that `.counter` is useful -- it is how a `backoff_counter_t` value is converted to a plain `uint16_t` to store in certain (more) public spots.

markshannon replied:
I still think that dropping the `counter` here and allowing `backoff_counter_t` in more "public spots" would make the code considerably simpler.

markshannon commented on Include/internal/pycore_backoff.h:
```diff
@@ -0,0 +1,117 @@
+}
+
+static inline backoff_counter_t
+forge_backoff_counter(uint16_t counter)
```
Do you mean "force"?

gvanrossum replied:
I meant forge -- it's a synonym for make. :-)

markshannon replied:
Having two different ways to make a backoff_counter seems confusing. It is not clear which one is the correct one to use.

markshannon commented on Include/internal/pycore_backoff.h:
```diff
@@ -0,0 +1,117 @@
+}
+
+static inline backoff_counter_t
+increment_backoff_counter(backoff_counter_t counter)
```
~Why do we need both increment *and* decrement?~ (Because of INSTRUMENTED_INSTRUCTION).

Once the counter is set, it should just "advance". Whether that is up or down should be an implementation detail.

markshannon commented on Include/internal/pycore_backoff.h:
```diff
@@ -0,0 +1,117 @@
+}
+
+static inline bool
+backoff_counter_is_zero(backoff_counter_t counter)
```
To avoid implementation details, maybe just `counter_triggers`?
Or merge into the "advance" which could return a bool indicating whether it has reached the threshold:
```C
bool backoff_counter_advance(backoff_counter_t *counter)
```

markshannon commented on Include/internal/pycore_code.h:
```diff
@@ -477,13 +473,9 @@ write_location_entry_start(uint8_t *ptr, int code, int length)
-
-
 static inline uint16_t
 adaptive_counter_bits(uint16_t value, uint16_t backoff) {
```
Maybe drop the `adaptive_counter` functions and replace them with the `backoff_counter` equivalents?

To do that cleanly we will probably need to change `_Py_CODEUNIT` to:
```C
typedef union {
    uint16_t cache;
    struct {
        uint8_t code;
        uint8_t arg;
    } op;
    backoff_counter_t counter;
} _Py_CODEUNIT;
```

gvanrossum replied:
I decided to keep them and not expose the backoff counter structure to other places.

gvanrossum replied:
I could dispense with `adaptive_counter_bits()`, using `make_backoff_counter()` directly below, but I would oppose getting rid of the `cooldown()` and `warmup()` helpers, because they are used in several/many places.

markshannon commented on Include/internal/pycore_interp.h:
```diff
@@ -239,10 +239,8 @@ struct _is {
 
-    /* These two values are shifted and offset to speed up check in JUMP_BACKWARD */
-    uint32_t optimizer_resume_threshold;
+    /* These three values are shifted and offset to speed up check in JUMP_BACKWARD */
```
`optimizer_resume_threshold` has been deleted, so it is now only two values.

markshannon commented on Python/bytecodes.c:
```diff
@@ -2348,41 +2349,35 @@ dummy_func(
-                    assert(tstate->previous_executor == NULL);
-                    tstate->previous_executor = Py_None;
-                    GOTO_TIER_TWO(executor);
+            if (ADAPTIVE_COUNTER_IS_ZERO(counter)) {
```
We could initialize this in `_PyCode_Quicken`, which would save the extra branch here.

gvanrossum replied:
That's tricky, because we can turn different optimizers off dynamically. The default (dummy) optimizer wants to initialize this to UNREACHABLE_BACKOFF, the uops optimizer uses value=16, and the counter optimizer (used in tests) uses value=0. The old strategy was more flexible, at least until the backoff kicked in. My current strategy is medium in flexibility -- the backoff sequence is initialized when the function first reaches the JUMP_BACKWARD. I've tried to minimize the cost of the extra branch by putting the ADAPTIVE_COUNTER_IS_ZERO() check before it, so it only hits when we're about to create a trace, which surely overwhelms the cost of an extra branch.

markshannon commented on Python/bytecodes.c:
```diff
@@ -4159,34 +4159,36 @@ dummy_func(
-                        Py_DECREF(previous);
-                        tstate->previous_executor = Py_None;
-                        GOTO_UNWIND();
+            #if ENABLE_SPECIALIZATION
```
This is tier 2 only, so making this conditional on `ENABLE_SPECIALIZATION` won't have any effect.
if `ENABLE_SPECIALIZATION` is false, then `_COLD_EXIT` will never be executed.

gvanrossum replied:
Yeah, I just realized this should not be enclosed in `#if ENABLE_SPECIALIZATION`.

(Thanks for the review -- I will work on this later.)

markshannon commented on Python/ceval_macros.h:
```diff
@@ -291,30 +291,32 @@ GETITEM(PyObject *v, Py_ssize_t i) {
+
 #else
+
 #define DECREMENT_ADAPTIVE_COUNTER(COUNTER)           \
```
Replace this with the backoff counter function?

gvanrossum replied:
I am hoping to avoid 30+ diff chunks where I just replace this macro call with another. Also it is (for now) redefined when the GIL is disabled.

markshannon replied:
I wouldn't worry about a few extra lines in the diff (there are 14 uses of `DECREMENT_ADAPTIVE_COUNTER` in bytecodes.c) if it makes the code simpler. I only need to review the diff once or twice, we need to work with the merged code for years.

markshannon commented on Python/ceval_macros.h:
```diff
@@ -291,30 +291,32 @@ GETITEM(PyObject *v, Py_ssize_t i) {
     } while (0);
-#endif
 
 #define INCREMENT_ADAPTIVE_COUNTER(COUNTER)          \
```
Likewise

gvanrossum replied:
This one I changed, it's much more esoteric. It's now called PAUSE.

markshannon reviewed (commented):
Looks promising. This should clarify the code a lot, especially if we replace all the `ADAPTIVE_` macros with `backoff_counter` functions.

You might also want to consider adding a `backoff_counter_t counter` field to the `_Py_CODEUNIT` union to avoid having to cast to and from `uint16_t` when accessing the counter.

gvanrossum committed:
Revert "In no-GIL mode, make INCREMENT_ADAPTIVE_COUNTER() a no-op"

I have a better way.

This reverts commit f6bf19499021fd26101e88111ae7d5b032c5ea25.

gvanrossum committed:
Fix comment and fix size of optimizer_backedge_threshold

gvanrossum committed:
_COLD_EXIT is not conditional on ENABLE_SPECIALIZATION

gvanrossum committed:
Rewrite _COLD_EXIT using backoff_counter_t directly

This also reverts the extra indent for the code I didn't touch.

gvanrossum committed:
Rename increment,decrement to advance,pause

Pause does `counter.value += 1` without any asserts.
I left is_zero alone; what to call it? is_ripe? :-)
There shouldn't be a need to check for POP_JUMP_IF,
it could affect perf a bit but with instrumentation
it's terrible anyways.

gvanrossum committed:
Give up on restricting backoff to <= 12

gvanrossum committed:
Rip out backoff thresholds

gvanrossum committed:
Fix tests

gvanrossum committed:
Fix initial temperature, clean up

gvanrossum committed:
Merge branch 'main' into exp-backoff

gvanrossum committed:
Remove unused variable

gvanrossum committed:
Admit defeat: advance_backoff_counter() may be entered when value == 0

This caused test_deadlock() in test_importlib to occasionally crash.

gvanrossum committed:
Merge branch 'main' into exp-backoff

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into exp-backoff

gvanrossum committed:
Put backoff field before value

markshannon commented on Include/internal/pycore_backoff.h:
```diff
@@ -0,0 +1,113 @@
+    };
+} backoff_counter_t;
+
+static_assert(sizeof(backoff_counter_t) == 2, "backoff counter size should be 2 bytes");
```
```suggestion
static_assert(sizeof(backoff_counter_t) == sizeof(_Py_CODEUNIT), "backoff counter size should be the same size as a code unit");
```

markshannon commented on Include/internal/pycore_backoff.h:
```diff
@@ -0,0 +1,113 @@
+}
+
+static inline backoff_counter_t
+reset_backoff_counter(backoff_counter_t counter)
```
This used to restart the counter, increasing the backoff, rather than resetting it? 
Maybe `restart_counter_and_backoff`?

gvanrossum replied:
Yes, it increments backoff and restarts the counter at `2**(backoff)`. I renamed it to `restart_backoff_counter`. (All the APIs here have `backoff_counter` in their name, to distinguish them from other forms of counters.)

markshannon commented on Include/internal/pycore_code.h:
```diff
@@ -451,18 +451,14 @@ write_location_entry_start(uint8_t *ptr, int code, int length)
- * counter set to (2**backoff - 1).
- * Backoff == 6 -> starting counter == 63, backoff == 10 -> starting counter == 1023.
+ * See pycore_backoff.h for more details.
+ * On a specialization failure, the backoff counter is reset.
```
```suggestion
 * On a specialization failure, the backoff counter is restarted.
```

markshannon commented on Include/internal/pycore_code.h:
```diff
@@ -503,13 +495,7 @@ adaptive_counter_cooldown(void) {
-    }
-    uint16_t value = (uint16_t)(1 << backoff) - 1;
-    return adaptive_counter_bits(value, backoff);
+    return reset_backoff_counter(forge_backoff_counter(counter)).counter;
```
I find this confusing. Is there a way to do this with a single call, rather than two calls and a field read?

markshannon commented on Lib/test/test_capi/test_opt.py:
```diff
@@ -69,7 +69,8 @@ def loop():
                     loop()
-                self.assertEqual(opt.get_count(), 1000)
+                # Subtract 16 because optimizer doesn't kick in until 16
+                self.assertEqual(opt.get_count(), 1000 - 16)
```
```suggestion
                self.assertLessEqual(opt.get_count(), 1000 - _testinternalcapi.TIER2_THRESHOLD)
```
We are bound to change the threshold, so it would be good not to hard code it in the tests.

markshannon commented on Lib/test/test_capi/test_opt.py:
```diff
@@ -96,7 +97,7 @@ def long_loop():
             self.assertEqual(opt.get_count(), 0)
             long_loop()
-            self.assertEqual(opt.get_count(), 10)
+            self.assertEqual(opt.get_count(), 20 - 16)  # Need 16 iterations to warm up
```
See above comment.

gvanrossum committed:
Small cleanup in .h files

gvanrossum committed:
Rename DECREMENT_ADAPTIVE_COUNTER to ADVANCE_...

gvanrossum committed:
Rename ADAPTIVE_COUNTER_IS_ZERO to ..._TRIGGERS

gvanrossum committed:
Rename backoff_counter_is_zero to ..._triggers

gvanrossum committed:
Rename reset_background_counter to restart_...

gvanrossum committed:
Make _Py_BackoffCounter a member of _Py_CODEUNIT

This changes a lot of things but the end result is arguably better.

gvanrossum committed:
Refactor initial counter values.

- The initial exit temperature is 64; this must be greater
  than the specialization cooldown value (52) otherwise we might
  create a trace before we have re-specialized the Tier 1 bytecode
- There's now a handy helper function for every counter initialization

gvanrossum committed:
Export tier 2 threshold from _testinternalcapi

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into exp-backoff

gvanrossum committed:
Add news

gvanrossum commented on Include/cpython/optimizer.h:
```diff
@@ -89,7 +89,7 @@ static inline uint16_t uop_get_error_target(const _PyUOpInstruction *inst)
 typedef struct _exit_data {
     uint32_t target;
-    int16_t temperature;
+    _Py_BackoffCounter temperature;
```
`temperature` is now a bit of a misnomer, since it counts down. Maybe it should be renamed to `counter` (same as in `CODEUNIT`)?

markshannon replied:
I'm fine with either.

gvanrossum replied:
I'll keep `temperature`, despite the misnomer -- it stands out and makes it easy to grep for this particular concept.

gvanrossum commented on Include/internal/pycore_backoff.h:
```diff
@@ -95,16 +83,43 @@ advance_backoff_counter(backoff_counter_t counter)
 
-static inline uint16_t
-initial_backoff_counter(void)
+/* Initial JUMP_BACKWARD counter.
```
There's a lot of boilerplate for these initial values; I followed your lead for `adaptive_counter_warmup()` and `adaptive_counter_cooldown()`, more or less.

markshannon reviewed (approved):
Looks good. Thanks for doing this.

gvanrossum committed:
Fix blurb formatting (I hope)
